### DISCUSSION TITLE
lit-html SSR templateResult causes "Error: invalid template strings array"

### DISCUSSION THREAD

**Discussion by unonweb** [lit/lit#3640] [Q&A] [answered]
Hi,

I'm trying to use lit-html for a minimalistic server side pre-production of html in **nodejs**:
Like this:
```js
function renderMenuBar(block = {}) {
   return html`
      <un-menu-bar theme=${ifDefined(theme)}>
         ${renderBlocksArray(block.blocks)}
      </un-menu-bar>
   `
}
```
The resulting _templateResult_ is saved to a js module and then imported like
```js
import ssrResult from './ssr-result'
```
But when lit tries to render the templateResult I get:
`Error: invalid template strings array`
Digging a bit deeper I find the dev_mode error message:

`Internal Error: expected template strings to be an array with a 'raw' field. Faking a template strings array by calling html or svg like an ordinary function is effectively the same as calling unsafeHtml and can lead to major security issues, e.g. opening your code up to XSS attacks.`

Ok, I see the `raw` field is missing when lit-html is executed in nodejs. 
My quesition is: Is it terribly wrong to use lit-html in node? 
If not - how can I prepare templateResults server side that can be used client side?
Do I need the whole SSR package for that (little bit intimidating)?

**Comment by augustjk** [comment, marked as answer]
Are you trying to send the template result over the network? That won't work as it loses its shape during serialization. If you want to do that, you would need to re-create the shape on the other side of the network but as the error message states this is a security risk since the template result coming in over the network could be untrustworthy.

`lit-html` can work in the server to produce a template result (and wouldn't just remove the `raw` field when consumed on the server) that can be rendered by `@lit-labs/ssr`, which is what you need to generate HTML strings from a Lit template.

**Reply by unonweb** [reply]
I tried to save the templateResult as json, so yes it was serialized. When re-creating the shape with `JSON.parse(templateResult)` lit's render function complained about the missing raw field. Indeed, when examining the resulting object it was the only difference. But when I added it manually just to see if I can trick the render function it still not worked (maybe because it somehow didn't match with the inline html comments that lit introduces?). 
Well, thanks for your reply. I will try lit/ssr now.
